src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1287 @@
+
+            layerOffset += layerAttributeCount;
+            arrayAppend(layerSizes, layerOffset);
+        }
+    }
+
+    /* Convert back to the default deleter to avoid dangling deleter function
+       pointer issues when unloading the plugin */
+    arrayShrink(flatAttributes, DefaultInit);
+    arrayShrink(layerSizes, DefaultInit);
+
+    return MaterialData{types, std::move(flatAttributes), std::move(layerSizes)};
+}
+
+UnsignedInt UfbxImporter::doTextureCount() const {
+    return UnsignedInt(_state->textures.size());
+}
+
+Int UfbxImporter::doTextureForName(Containers::StringView name) {
+    const ufbx_element* element = ufbx_find_element_len(_state->scene.get(), UFBX_ELEMENT_TEXTURE, name.data(), name.size());
+    return element ? _state->textureRemap[element->typed_id] : -1;
+}
+
+Containers::String UfbxImporter::doTextureName(UnsignedInt id) {
+    const FileTexture& fileTexture = _state->textures[id];
+    return _state->scene->textures[fileTexture.textureIndex]->name;
+}
+
+Containers::Optional<TextureData> UfbxImporter::doTexture(UnsignedInt id) {
+    const FileTexture& fileTexture = _state->textures[id];
+    const ufbx_texture* texture = _state->scene->textures[fileTexture.textureIndex];
+
+    const SamplerWrapping wrappingU = toSamplerWrapping(texture->wrap_u);
+    const SamplerWrapping wrappingV = toSamplerWrapping(texture->wrap_v);
+    return TextureData{TextureType::Texture2D,
+        SamplerFilter::Linear, SamplerFilter::Linear, SamplerMipmap::Linear,
+        {wrappingU, wrappingV, SamplerWrapping::ClampToEdge},
+        fileTexture.fileTextureIndex};
+}
+
+AbstractImporter* UfbxImporter::setupOrReuseImporterForImage(UnsignedInt id, const char* errorPrefix) {
+    const ufbx_texture_file& file = _state->scene->texture_files[id];
+
+    /* Looking for the same ID, so reuse an importer populated before. If the
+       previous attempt failed, the importer is not set, so return nullptr in
+       that case. Going through everything below again would not change the
+       outcome anyway, only spam the output with redundant messages. */
+    if(_state->imageImporterId == id)
+        return _state->imageImporter ? &*_state->imageImporter : nullptr;
+
+    /* Otherwise reset the importer and remember the new ID. If the import
+       fails, the importer will stay unset, but the ID will be updated so the
+       next round can again just return nullptr above instead of going through
+       the doomed-to-fail process again. */
+    _state->imageImporter = Containers::NullOpt;
+    _state->imageImporterId = id;
+
+    AnyImageImporter importer{*manager()};
+    importer.setFlags(flags());
+    if(fileCallback()) importer.setFileCallback(fileCallback(), fileCallbackUserData());
+
+    if(file.content.size > 0) {
+        auto textureData = Containers::ArrayView<const char>(reinterpret_cast<const char*>(file.content.data), file.content.size);
+        if(!importer.openData(textureData))
+            return nullptr;
```

Comment:
Codecov says that everything is covered, except this `return` line, even though it should be tested with `imageBrokenEmbedded()`. See my comment there, I might have an idea :)

Nevertheless, huge congrats for the 99.86% code coverage, unbelievable achievement :exploding_head:  Lots of technical capital (as opposed to debt) here :blush:

Reply:
Yea I was wondering what's up with not getting to 100% even though I tried to fish for it, for me it looked like every line was covered in Codecov but probably skimmed over it. And thanks :D please don't enable branch coverage and ruin my perfectly green file, doing that for ufbx went from 95% -> 87% coverage.

Reply:
> please don't enable branch coverage

No worries, I'm still too scared to enable that even for local builds :D

Reply:
Thanks for the tip about comparing the error message, turns out the failure happened at `image2D()` and not at `openData()` as the error happened in PNG parsing. I modified the test file so that it has an unknown magic number and got that line covered. Yay 100%!

![image](https://user-images.githubusercontent.com/3875461/208268537-9e005bf4-e75f-4337-9c55-43e48c24a8df.png)